mysql_mimic/results.py

```diff
@@ -0,0 +1,237 @@
+
+
+def _bencode_bytes(val):
+    return str_len(val)
+
+
+def _bencode_date(val):
+    year = val.year
+    month = val.month
+    day = val.day
+
+    if isinstance(val, datetime):
+        hour = val.hour
+        minute = val.minute
+        second = val.second
+        microsecond = val.microsecond
+    else:
+        hour = minute = second = microsecond = 0
+
+    if microsecond == 0:
+        if hour == minute == second == 0:
+            if year == month == day == 0:
+                return uint_1(0)
+            return b"".join([uint_1(4), uint_2(year), uint_1(month), uint_1(day)])
+        return b"".join(
+            [
+                uint_1(7),
+                uint_2(year),
+                uint_1(month),
+                uint_1(day),
+                uint_1(hour),
+                uint_1(minute),
+                uint_1(second),
+            ]
+        )
+    return b"".join(
+        [
+            uint_1(11),
+            uint_2(year),
+            uint_1(month),
+            uint_1(day),
+            uint_1(hour),
+            uint_1(minute),
+            uint_1(second),
+            uint_4(microsecond),
+        ]
+    )
+
+
+def _bencode_int(val):
+    return struct.pack("<q", val)
+
+
+def _bencode_float(val):
+    return struct.pack("<d", val)
+
+
+def _bencode_timedelta(val):
+    days = abs(val.days)
+    hours, remainder = divmod(abs(val.seconds), 3600)
+    minutes, seconds = divmod(remainder, 60)
+    microseconds = val.microseconds
+    is_negative = val.total_seconds() < 0
+
+    if microseconds == 0:
+        if days == hours == minutes == seconds == 0:
+            return uint_1(0)
+        return b"".join(
+            [
+                uint_1(8),
+                uint_1(is_negative),
+                uint_4(days),
+                uint_1(hours),
+                uint_1(minutes),
+                uint_1(seconds),
+            ]
+        )
+    return b"".join(
+        [
+            uint_1(12),
+            uint_1(is_negative),
+            uint_4(days),
+            uint_1(hours),
+            uint_1(minutes),
+            uint_1(seconds),
+            uint_4(microseconds),
+        ]
+    )
+
+
+# Order matters
+# bool is a subclass of int
+# datetime is a subclass of date
+_ENCODERS = {
+    # python type: (mysql type, text encoder, binary encoder),
+    bool: (ColumnType.TINY, lambda v: str(int(v)), _bencode_bool),
+    datetime: (ColumnType.DATETIME, str, _bencode_date),
+    str: (ColumnType.STRING, lambda v: v, _bencode_str),
+    bytes: (ColumnType.BLOB, lambda v: v, _bencode_bytes),
+    int: (ColumnType.LONGLONG, str, _bencode_int),
+    float: (ColumnType.DOUBLE, str, _bencode_float),
+    date: (ColumnType.DATE, str, _bencode_date),
+    timedelta: (ColumnType.TIME, str, _bencode_timedelta),
+}
+
+
+def bencode(val):
+    for py_type, (_, _, encoder) in _ENCODERS.items():
+        if isinstance(val, py_type):
+            return encoder(val)
+    return _bencode_str(val)
+
+
+def tencode(val):
```

Comment:
bencode is byte encode, tencode is ?

Reply:
text encode.
MySQL has "Text result sets" and "binary result sets". So bencode is binary encode.
Let me add more explicit names